### DISCUSSION TITLE
rewrites should set proxy headers

### DISCUSSION THREAD

**Discussion by lukasz-karolewski** [vercel/next.js#17325] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

I have separate api server, which also my auth server. I'm using cookie authentication and oauth2.

In prod I have a caddy server in front, which is a transparent proxy (preserving Host header) - everything works ok.

On local machine I'm unable to proxy dev server hot reload via caddy, and with nextjs rewrites my auth is not working because of missing proxy headers.

## Describe the solution you'd like

rewrites should set proxy headers, at minimum X_FORWARDED_HOST and X_FORWARDED_PORT,

alternatively act like a transparent proxy by preserving Host header

**Comment by jacargentina** [comment]
Yes, I need exactly the same.

I should note X-FORWARDED-FOR too!

**Comment by javiermatos** [comment]
I am having the exact same issue. I use nginx as reverse proxy with pages served using nextjs, but api processed by Django.

When I request a page, nextjs will do the server side rendering and call the API, which is the Django application. When Django has to generate the urls for media images, it assumes server address is "localhost", so images will use localhost as domain.

Is there any way to have fetch in server side rendering to forward those headers?

**Reply by pfcodes** [reply]
I'm in the same boat, did you figure out a solution?

**Comment by thomasmarshall** [comment]
I opened a PR for this same issue: https://github.com/vercel/next.js/pull/17557

In the meantime, you might be interested in using a custom proxy API route as a temporary solution:

```js
// next.config.js

module.exports = {
  rewrites: async () => [
    {
      source: '/:path*',
      destination: '/:path*',
    },
    {
      source: '/:path*',
      destination: '/api/proxy/:path*',
    },
  ],
};
```

```js
// pages/api/proxy/[[...path]].js

import { createProxyMiddleware } from 'http-proxy-middleware';

export default createProxyMiddleware({
  target: process.env.PROXY_HOST,
  changeOrigin: true,
  pathRewrite: { '^/api': '/' },
  xfwd: true,
});

export const config = {
  api: {
    bodyParser: false, // enable POST requests
    externalResolver: true, // hide warning message
  },
};
```

**Reply by migom6** [reply]
I also have a same setup but writing a custom proxy inside rewrites messed up a lot of things for me, I hope they make rewrites as transparent as possible.

**Reply by rvetere** [reply]
Works, thanks guys! But i wonder... on every single request, we see another "Proxy created: / -> xyz".. after a few requests, we always get an error:

`MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 close listeners added to [Server]. Use emitter.setMaxListeners() to increase limit`

If we change the proxy implementation into a singleton, the proxy just won't work anymore.... does anybody of you know where this is coming from and how to fix this?

**Reply by nemo369** [reply]
Thank you so much for this
